File: JeMPI_Apps/JeMPI_AsyncReceiver/src/main/java/org/jembi/jempi/async_receiver/CustomAsyncHelper.java
package org.jembi.jempi.async_receiver;

import org.apache.commons.csv.CSVRecord;
import org.jembi.jempi.shared.models.CustomDemographicData;
import org.jembi.jempi.shared.models.CustomSourceId;
import org.jembi.jempi.shared.models.CustomUniqueInteractionData;

final class CustomAsyncHelper {

   private static final int AUX_ID_COL_NUM = 0;
   private static final int AUX_CLINICAL_DATA_COL_NUM = 15;
   private static final int SOURCEID_FACILITY_COL_NUM = 13;
   private static final int SOURCEID_PATIENT_COL_NUM = 14;
   private static final int GIVEN_NAME_COL_NUM = 1;
   private static final int FAMILY_NAME_COL_NUM = 2;
   private static final int GENDER_COL_NUM = 3;
   private static final int DOB_COL_NUM = 4;
   private static final int CITY_COL_NUM = 5;
   private static final int PHONE_NUMBER_HOME_COL_NUM = 6;
   private static final int PHONE_NUMBER_MOBILE_COL_NUM = 7;
   private static final int PHN_COL_NUM = 8;
   private static final int NIC_COL_NUM = 9;
   private static final int PPN_COL_NUM = 10;
   private static final int SCN_COL_NUM = 11;
   private static final int DL_COL_NUM = 12;

   private CustomAsyncHelper() {
   }

   static CustomUniqueInteractionData customUniqueInteractionData(final CSVRecord csvRecord) {
      return new CustomUniqueInteractionData(java.time.LocalDateTime.now(),
                                             Main.parseRecordNumber(csvRecord.get(AUX_ID_COL_NUM)),
                                             csvRecord.get(AUX_CLINICAL_DATA_COL_NUM));
   }

   static CustomDemographicData customDemographicData(final CSVRecord csvRecord) {
      return new CustomDemographicData(
         csvRecord.get(GIVEN_NAME_COL_NUM),
         csvRecord.get(FAMILY_NAME_COL_NUM),
         csvRecord.get(GENDER_COL_NUM),
         csvRecord.get(DOB_COL_NUM),
         csvRecord.get(CITY_COL_NUM),
         csvRecord.get(PHONE_NUMBER_HOME_COL_NUM),
         csvRecord.get(PHONE_NUMBER_MOBILE_COL_NUM),
         csvRecord.get(PHN_COL_NUM),
         csvRecord.get(NIC_COL_NUM),
         csvRecord.get(PPN_COL_NUM),
         csvRecord.get(SCN_COL_NUM),
         csvRecord.get(DL_COL_NUM));
   }

   static CustomSourceId customSourceId(final CSVRecord csvRecord) {
      return new CustomSourceId(
         null,
         csvRecord.get(SOURCEID_FACILITY_COL_NUM),
         csvRecord.get(SOURCEID_PATIENT_COL_NUM));
   }

}

File: JeMPI_Apps/JeMPI_Configuration/src/main/scala/configuration/CustomAsyncHelper.scala
package configuration

import java.io.{File, PrintWriter}

private object CustomAsyncHelper {

  private val classLocation =
    "../JeMPI_AsyncReceiver/src/main/java/org/jembi/jempi/async_receiver"
  private val customClassName = "CustomAsyncHelper"
  private val packageText = "org.jembi.jempi.async_receiver"

  def generate(config: Config): Unit =
    val classFile: String =
      classLocation + File.separator + customClassName + ".java"
    println("Creating " + classFile)
    val file: File = new File(classFile)
    val writer: PrintWriter = new PrintWriter(file)
    writer.println(s"""package $packageText;
         |
         |import org.apache.commons.csv.CSVRecord;
         |import org.jembi.jempi.shared.models.CustomSourceId;
         |import org.jembi.jempi.shared.models.CustomUniqueInteractionData;
         |
         |final class $customClassName {
         |
         |${columnIndices()}
         |
         |   private $customClassName() {
         |   }
         |
         |   static CustomUniqueInteractionData customUniqueInteractionData(final CSVRecord csvRecord) {
         |      return new CustomUniqueInteractionData(${customUniqueInteractionArguments()});
         |   }
         |
         |${
                       if (config.additionalNodes.isEmpty) ""
                       else
                         config.additionalNodes.get
                           .map(x => customNodeConstructor(x))
                           .mkString
                     }
         |}
         |""".stripMargin)

    writer.flush()
    writer.close()

    def columnIndices(): String =

      def additionalNodeFields(additionalNode: AdditionalNode): String =
        additionalNode.fields
          .map(f =>
            if (f.csvCol.isEmpty) ""
            else
              s"""   private static final int ${additionalNode.nodeName.toUpperCase}_${f.fieldName.toUpperCase}_COL_NUM = ${f.csvCol.get};
               |""".stripMargin
          )
          .mkString("")
          .stripMargin
          .stripTrailing()
      end additionalNodeFields

      def uniqueInteractionFields(fields: Array[UniqueField]): String =
        fields
          .map(f =>
            if (f.csvCol.isEmpty) ""
            else
              s"""   private static final int ${f.fieldName
                  .toUpperCase()}_COL_NUM = ${f.csvCol.get};
               |""".stripMargin
          )
          .mkString("")
          .stripMargin
      end uniqueInteractionFields

      (if (config.uniqueInteractionFields.isEmpty) ""
       else
         uniqueInteractionFields(config.uniqueInteractionFields.get))
        +
          (if (
             config.additionalNodes.isEmpty || config.additionalNodes.get.isEmpty
           ) ""
           else
             config.additionalNodes.get
               .map(x => s"""${additionalNodeFields(x)}""")
               .mkString(sys.props("line.separator"))) + sys.props(
            "line.separator"
          )
          +
          config.demographicFields
            .filter(f => f.source.isDefined && f.source.get.csvCol.isDefined)
            .map(f =>
              s"""${" " * 3}private static final int ${f.fieldName.toUpperCase}_COL_NUM = ${f.source.get.csvCol.get};"""
            )
            .mkString(sys.props("line.separator"))
    end columnIndices

    def demographicFields(): String =
      config.demographicFields
        .map(f =>
          if (f.source.isDefined && f.source.get.generate.isDefined) {
            s"""${" " * 9}null,"""
          } else {
            s"""${" " * 9}csvRecord.get(${f.fieldName.toUpperCase}_COL_NUM),"""
          }
        )
        .mkString(sys.props("line.separator"))
        .dropRight(1)
    end demographicFields

    def customUniqueInteractionArguments(): String =
      if (config.uniqueInteractionFields.isEmpty) ""
      else
        config.uniqueInteractionFields.get
          .map(f =>
            if (f.fieldName.toUpperCase.equals("AUX_ID")) {
              s"""${" " * 45}Main.parseRecordNumber(csvRecord.get(${f.fieldName.toUpperCase}_COL_NUM))"""
            } else if (f.fieldName.toUpperCase.equals("AUX_DATE_CREATED")) {
              s"""${" " * 45}java.time.LocalDateTime.now()"""
            } else {
              s"""${" " * 45}csvRecord.get(${f.fieldName.toUpperCase}_COL_NUM)"""
            }
          )
          .mkString(""",
              |""".stripMargin)
          .trim
    end customUniqueInteractionArguments

    def customNodeConstructor(additionalNode: AdditionalNode): String =

      def arguments(fields: Array[AdditionalNodeField]): String =
        fields
          .map(f =>
            if (f.csvCol.isEmpty)
              s"""${" " * 9}null"""
            else
              s"""${" " * 9}csvRecord.get(${additionalNode.nodeName.toUpperCase}_${f.fieldName
                  .toUpperCase()}_COL_NUM)"""
          )
          .mkString(s",${sys.props("line.separator")}")
      end arguments

      s"""   static Custom${additionalNode.nodeName} custom${additionalNode.nodeName}(final CSVRecord csvRecord) {
         |      return new Custom${additionalNode.nodeName}(
         |         null,
         |${arguments(additionalNode.fields)});
         |   }
         |""".stripMargin
    end customNodeConstructor

  end generate

}
